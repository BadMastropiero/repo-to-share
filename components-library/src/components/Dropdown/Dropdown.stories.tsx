import {ComponentMeta, ComponentStory} from '@storybook/react';
import {useState} from 'react';

import Dropdown from './Dropdown';

import testingOptions from './optionsTesting';

export default {
  title: 'Components/Dropdown',
  component: Dropdown,
} as ComponentMeta<typeof Dropdown>;

export const Base: ComponentStory<typeof Dropdown> = args => {
  const {
    primary,
    disabled,
    placeholder,
    text,
    isMulti,
    isLoading,
    isClearable,
    isRtl,
    isSearchable,
    placeholderOut,
    renderSelected,
  } = args;
  const [selected, setSelected] = useState<boolean>(false);

  return (
    <Dropdown
      primary={primary}
      disabled={disabled}
      placeholder={placeholder}
      onClick={() => {
        setSelected(!selected);
      }}
      isMulti={isMulti}
      text={text}
      options={testingOptions}
      defaultValue={[testingOptions[0], testingOptions[1]]}
      isLoading={isLoading}
      isClearable={isClearable}
      isRtl={isRtl}
      isSearchable={isSearchable}
      placeholderOut={placeholderOut}
      renderSelected={renderSelected}
    />
  );
};

Base.args = {
  primary: true,
  disabled: false,
  placeholder: 'Select...',
  isMulti: true,
  isLoading: false,
  isClearable: undefined,
  isRtl: undefined,
  isSearchable: undefined,
  placeholderOut: false,
  renderSelected: true,
};

Base.parameters = {
  design: {
    type: 'figma',
    // url: 'https://www.figma.com/file/epHXucNHjke2rES69iOsaS/TetraImpacts?node-id=2%3A1562',
  },
};
