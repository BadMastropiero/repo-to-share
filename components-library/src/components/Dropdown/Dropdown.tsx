import React, {useContext} from 'react';
import {ThemeContext} from 'styled-components';

import Select, {
  components,
  DropdownIndicatorProps,
  IndicatorSeparatorProps,
  OptionProps,
  GroupBase,
} from 'react-select';

// import makeAnimated from 'react-select/animated';
import DropdownProps, {TiOptions} from './Dropdown.interface';

import {Button} from '../Button';
import {ArrowDown, ArrowUp} from '../Icon';
import {Option} from '../Option';
import {getStyles, StyledDropdown} from './Dropdown.styles';
import {Typography} from '../Typography';

/**
 * The following snippet is required in order to pass our custom props
 * to the select returned by our Dropdown, as described in
 * [this comment](https://github.com/JedWatson/react-select/issues/4804#issuecomment-927223471)
 */
declare module 'react-select/dist/declarations/src/Select' {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  export interface Props<Option, IsMulti extends boolean, Group extends GroupBase<Option>> {
    /**
     * Here we declare our custom props, which will be passed to the select
     * @param primary - Set the theme type.
     * @param optionWithCheckBox - If checkboxes are shown in options that populate the list.
     */
    primary?: boolean;
    optionWithCheckBox?: boolean;
  }
}

// TODO Maybe use this to animate the multivalues when closing
// const animatedComponents = makeAnimated();

const TiDropdownIndicator = ({...props}: DropdownIndicatorProps<TiOptions, true>) => (
  <components.DropdownIndicator {...props}>
    <Button
      beforeIcon={props.selectProps.menuIsOpen ? ArrowUp : ArrowDown}
      presetSize="small"
      primary={props.selectProps.primary}
      disabled={props.isDisabled}
      rounded
      text=""
    />
  </components.DropdownIndicator>
);

const TiIndicatorSeparator = ({innerProps}: IndicatorSeparatorProps<TiOptions, true>) => (
  <span {...innerProps} />
);

const TiOption = ({...props}: OptionProps<TiOptions, true>) => (
  <components.Option {...props}>
    <Option
      primary={props.selectProps.primary}
      text={props.label}
      selected={props.isSelected}
      noCheckbox={!props.selectProps.optionWithCheckBox}
    />
  </components.Option>
);

const Dropdown: React.ComponentType<DropdownProps> = ({
  name = 'ti-dropdown',
  disabled = false,
  primary = false,
  isMulti,
  defaultValue,
  isLoading = false,
  isClearable = false,
  isRtl = false,
  isSearchable = false,
  placeholder = 'Select..',
  options,
  menuIsOpen,
  closeMenuOnSelect = !isMulti,
  placeholderOut = false,
  renderSelected = false,
  setValue,
  value,
}) => {
  const {colors, typography} = useContext(ThemeContext);

  const customStyles = getStyles({colors, typography});

  return (
    <StyledDropdown>
      {placeholderOut && ( // optional external placeholder
        <Typography
          colorPreset="disable"
          preset="label"
          variant="p"
          style={{minWidth: 'min-content'}}
        >
          {placeholder}
        </Typography>
      )}
      <Select
        closeMenuOnSelect={closeMenuOnSelect}
        components={{
          DropdownIndicator: TiDropdownIndicator,
          IndicatorSeparator: TiIndicatorSeparator,
          Option: TiOption,
        }}
        options={options}
        isDisabled={disabled}
        isMulti={isMulti}
        defaultValue={defaultValue}
        isLoading={isLoading}
        isRtl={isRtl}
        isSearchable={isSearchable}
        name={name}
        placeholder={placeholderOut ? '' : placeholder}
        styles={customStyles}
        onChange={newValue => {
          if (setValue) setValue(newValue);
        }}
        value={value}
        menuIsOpen={menuIsOpen}
        primary={primary}
        controlShouldRenderValue={renderSelected}
        optionWithCheckBox={!renderSelected}
        isClearable={isClearable}
        backspaceRemovesValue
        hideSelectedOptions={renderSelected}
      />
    </StyledDropdown>
  );
};

export default Dropdown;
