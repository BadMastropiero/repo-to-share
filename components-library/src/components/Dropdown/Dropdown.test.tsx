import {render, screen} from '@testing-library/react';
import {composeStories} from '@storybook/testing-react';
import '@testing-library/jest-dom';
import * as stories from './Dropdown.stories';

const {Base} = composeStories(stories);

describe('Dropdown tests', () => {
  test('should render', () => {
    render(<Base />);
    expect(screen.getByRole('combobox')).toBeInTheDocument();
  });

  test('should match snapshot', () => {
    const tree = render(<Base />);
    expect(tree).toMatchSnapshot();
  });
});
