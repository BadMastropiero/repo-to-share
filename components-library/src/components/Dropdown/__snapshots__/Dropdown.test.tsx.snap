// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-csuSiG tFIWU"
      >
        <div
          class=" css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-3-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class=" css-dv4xrf-control"
          >
            <div
              class=" css-g1d714-ValueContainer"
            >
              <div
                class="css-1t4u7w3-multiValue"
              >
                <div
                  class="css-bmw50s"
                >
                  Han Solo
                </div>
                <div
                  aria-label="Remove Han Solo"
                  class="css-i54qw6"
                  role="button"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="14"
                    viewBox="0 0 20 20"
                    width="14"
                  >
                    <path
                      d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="css-1t4u7w3-multiValue"
              >
                <div
                  class="css-bmw50s"
                >
                  Darth Vader
                </div>
                <div
                  aria-label="Remove Darth Vader"
                  class="css-i54qw6"
                  role="button"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="14"
                    viewBox="0 0 20 20"
                    width="14"
                  >
                    <path
                      d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                    />
                  </svg>
                </div>
              </div>
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="true"
                aria-readonly="true"
                class="css-mohuvp-dummyInput-DummyInput"
                id="react-select-3-input"
                inputmode="none"
                role="combobox"
                tabindex="0"
                value=""
              />
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span />
              <div
                aria-hidden="true"
                class=" css-x0kyhz-indicatorContainer"
              >
                <button
                  class="sc-bcXHqe sc-dkrFOg gkrZeY ti-button "
                >
                  <div
                    class="sc-gswNZR ezwamZ"
                  >
                    <svg
                      class="sc-jSUZER ldObNo ticons-arrow_down"
                      height="16px"
                      role="img"
                      viewBox="0 0 24 24"
                      width="16px"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M21.3625 8.8126L12.7625 17.4126C12.6292 17.5459 12.4832 17.6416 12.3245 17.6996C12.1665 17.7583 12.0042 17.7876 11.8375 17.7876C11.6709 17.7876 11.5085 17.7583 11.3505 17.6996C11.1919 17.6416 11.0459 17.5459 10.9125 17.4126L2.28752 8.7876C1.98752 8.4876 1.83752 8.12926 1.83752 7.7126C1.83752 7.29593 1.99586 6.92926 2.31252 6.6126C2.61252 6.3126 2.97919 6.1626 3.41252 6.1626C3.84586 6.1626 4.21252 6.3126 4.51252 6.6126L11.8375 13.9376L19.1875 6.5876C19.4875 6.2876 19.8459 6.1376 20.2625 6.1376C20.6792 6.1376 21.0459 6.29593 21.3625 6.6126C21.6625 6.9126 21.8125 7.27926 21.8125 7.7126C21.8125 8.14593 21.6625 8.5126 21.3625 8.8126Z"
                      />
                    </svg>
                  </div>
                  <div
                    class="sc-gKPRtg dOMrEj"
                    color="#C5CFDF"
                  />
                </button>
              </div>
            </div>
          </div>
          <div>
            <input
              name="ti-dropdown"
              type="hidden"
              value="han"
            />
            <input
              name="ti-dropdown"
              type="hidden"
              value="darth"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-csuSiG tFIWU"
    >
      <div
        class=" css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-3-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-dv4xrf-control"
        >
          <div
            class=" css-g1d714-ValueContainer"
          >
            <div
              class="css-1t4u7w3-multiValue"
            >
              <div
                class="css-bmw50s"
              >
                Han Solo
              </div>
              <div
                aria-label="Remove Han Solo"
                class="css-i54qw6"
                role="button"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="14"
                  viewBox="0 0 20 20"
                  width="14"
                >
                  <path
                    d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="css-1t4u7w3-multiValue"
            >
              <div
                class="css-bmw50s"
              >
                Darth Vader
              </div>
              <div
                aria-label="Remove Darth Vader"
                class="css-i54qw6"
                role="button"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="14"
                  viewBox="0 0 20 20"
                  width="14"
                >
                  <path
                    d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                  />
                </svg>
              </div>
            </div>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-haspopup="true"
              aria-readonly="true"
              class="css-mohuvp-dummyInput-DummyInput"
              id="react-select-3-input"
              inputmode="none"
              role="combobox"
              tabindex="0"
              value=""
            />
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span />
            <div
              aria-hidden="true"
              class=" css-x0kyhz-indicatorContainer"
            >
              <button
                class="sc-bcXHqe sc-dkrFOg gkrZeY ti-button "
              >
                <div
                  class="sc-gswNZR ezwamZ"
                >
                  <svg
                    class="sc-jSUZER ldObNo ticons-arrow_down"
                    height="16px"
                    role="img"
                    viewBox="0 0 24 24"
                    width="16px"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M21.3625 8.8126L12.7625 17.4126C12.6292 17.5459 12.4832 17.6416 12.3245 17.6996C12.1665 17.7583 12.0042 17.7876 11.8375 17.7876C11.6709 17.7876 11.5085 17.7583 11.3505 17.6996C11.1919 17.6416 11.0459 17.5459 10.9125 17.4126L2.28752 8.7876C1.98752 8.4876 1.83752 8.12926 1.83752 7.7126C1.83752 7.29593 1.99586 6.92926 2.31252 6.6126C2.61252 6.3126 2.97919 6.1626 3.41252 6.1626C3.84586 6.1626 4.21252 6.3126 4.51252 6.6126L11.8375 13.9376L19.1875 6.5876C19.4875 6.2876 19.8459 6.1376 20.2625 6.1376C20.6792 6.1376 21.0459 6.29593 21.3625 6.6126C21.6625 6.9126 21.8125 7.27926 21.8125 7.7126C21.8125 8.14593 21.6625 8.5126 21.3625 8.8126Z"
                    />
                  </svg>
                </div>
                <div
                  class="sc-gKPRtg dOMrEj"
                  color="#C5CFDF"
                />
              </button>
            </div>
          </div>
        </div>
        <div>
          <input
            name="ti-dropdown"
            type="hidden"
            value="han"
          />
          <input
            name="ti-dropdown"
            type="hidden"
            value="darth"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
